File: java/L600/LeetCode0684.java
package L600;

public class LeetCode0684 {
    private class UnionFind {
        private int[] ids, ranks;

        public UnionFind(int N) {
            ids = new int[N];
            ranks = new int[N];
            for (int i = 0;i < N;++i) {
                ids[i] = i;
                ranks[i] = 1;
            }
        }

        public int find(int id) {
            while (ids[id] != id) {
                id = ids[id];
            }
            return id;
        }

        public void union(int p, int q) {
            int pid = find(p);
            int qid = find(q);
            if (pid != qid) {
                if (ranks[pid] <= ranks[qid]) {
                    ids[pid] = qid;
                } else {
                    ids[qid] = pid;
                }
                if (ranks[pid] == ranks[qid]) {
                    ranks[qid] += 1;
                }
            }
        }

        public boolean isConnected(int p, int q) {
            return find(p) == find(q);
        }
    }

    public int[] findRedundantConnection(int[][] edges) {
        UnionFind uf = new UnionFind(edges.length + 1);
        for (int[] edge : edges) {
            if (!uf.isConnected(edge[0], edge[1])) {
                uf.union(edge[0], edge[1]);
            } else {
                return edge;
            }
        }
        return new int[0];
    }
}

File: cpp/L600/LeetCode0684.cc
#include <vector>
using std::vector;

struct UnionFind {
    int n;
    vector<int> ids, ranks;

    UnionFind(int n): n(n), ids(n), ranks(n) {
        for (int i = 0;i < n;++i) {
            ids[i] = i;
            ranks[i] = 1;
        }
    }

    int to_find(int id) {
        while (ids[id] != id) {
            id = ids[id];
        }
        return id;
    }

    bool to_union(int p, int q) {
        int pid = to_find(p);
        int qid = to_find(q);
        if (pid == qid) {
            return false;
        }
        if (ranks[pid] >= ranks[qid]) {
            ids[qid] = pid;
            ranks[pid] += ranks[qid];
        } else {
            ids[pid] = qid;
            ranks[qid] += ranks[pid];
        }
        return true;
    }
};

class Solution {
public:
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        UnionFind uf(edges.size());
        for (vector<int> const &edge : edges) {
            bool f = uf.to_union(edge[0] - 1, edge[1] - 1);
            if (!f) {
                return edge;
            }
        }
        return {};
    }
};